logstash-core/src/main/java/org/logstash/config/ir/compiler/DatasetCompiler.java

```diff
@@ -196,12 +200,13 @@ private static Closure filterBody(final ValueSyntaxElement outputBuffer,
         final ValueSyntaxElement inputBuffer, final ClassFields fields,
         final AbstractFilterDelegatorExt plugin) {
         final ValueSyntaxElement filterField = fields.add(plugin);
-        final Closure body = Closure.wrap(
+        final Closure body = Closure.wrap(setPluginIdForLog4j(plugin),
```

Comment:
Because method `filterDataset` invokes `filterBody` and filterBody sets and unsets the plugin.id in the context, while `outputDataset` create directly the statements creators, without invoking the filterBody. 
So in filterBody, I need to set to make working in filters while in outputDataset for output.